src/fs.rs

```diff
@@ -0,0 +1,449 @@
+}
+
+/// Use this function for updating a file based on its current content,
+/// for files which might be accessed at the same time by other threads/processes.
+/// The basic principle is to create a temporary file, take a lock on the file to be rewritten,
+/// do the rewrite (which might involve reading the old file contents),
+/// and finally release the lock.
+/// Error handling (especially the case where locking is not possible) makes the implementation
+/// non-straightforward.
+///
+/// # Arguments
+///
+/// - `path`: The path to the file which should be updated.
+/// - `locking_enabled`: To indicate whether locking should be attempted. This function might
+/// update the value stored in `locking_enabled`.
+/// - `rewrite`: The function which handles reading from the file and writing to a temporary file.
+/// The first argument is for the file to read from, the second for the temporary file to write to.
+/// On success, the value returned by `rewrite` is included in this functions return value. Be
+/// careful about side effects of `rewrite`. It might get executed multiple times. Try to avoid
+/// side effects and instead extract any data you might need and return them on success.
+/// Then, apply the desired side effects once this function has returned successfully.
+///
+/// # Return value
+///
+/// On success, the [`FileId`] of the rewritten file is returned, alongside the value returned by
+/// `rewrite`. Note that if locking is unavailable, the [`FileId`] might be the id of a different
+/// version of the file, which was written after this function renamed the temporary file to `path`
+/// but before we obtained the [`FileId`] from `path`. This is a race condition we do not detect.
+pub fn rewrite_via_temporary_file<F, U>(
+    path: &wstr,
+    locking_enabled: &RelaxedAtomicBool,
+    rewrite: F,
+) -> std::io::Result<(FileId, U)>
+where
+    F: Fn(&mut File, &mut File) -> std::io::Result<U>,
+{
+    /// Updates the metadata of the `new_file` to match the `old_file`.
+    /// Also updates the mtime of the `new_file` to the current time manually, to work around
+    /// operating systems which do not update the internal time used for such time stamps
+    /// frequently enough. This is important for [`FileId`] comparisons.
+    fn update_metadata(old_file: &File, new_file: &File) {
+        // Ensure we maintain the ownership and permissions of the original (#2355). If the
+        // stat fails, we assume (hope) our default permissions are correct. This
+        // corresponds to e.g. someone running sudo -E as the very first command. If they
+        // did, it would be tricky to set the permissions correctly. (bash doesn't get this
+        // case right either).
+        if let Ok(md) = old_file.metadata() {
+            // TODO(MSRV): Consider replacing with std::os::unix::fs::fchown when MSRV >= 1.73
+            if unsafe { fchown(new_file.as_raw_fd(), md.uid(), md.gid()) } == -1 {
```

Comment:
(not really relevant for this PR but I think this should be retried on EINTR.
Of course that's very unlikely to have caused issues. We'll get retries for free when we upgrade to `std::os::unix::fs::fchown` or another wrapper)